tensorboard/webapp/app_routing/effects/app_routing_effects_test.ts

```diff
@@ -517,6 +534,103 @@ describe('app_routing_effects', () => {
           }),
         ]);
       }));
+
+      describe('programmatical navigation integration', () => {
+        it('redirects without query parameter', fakeAsync(() => {
+          getPathSpy.and.returnValue('/redirect_no_query/compare');
+          getSearchSpy.and.returnValue([]);
+          deserializeQueryParamsSpy.and.returnValue({});
+
+          action.next(effects.ngrxOnInitEffects());
+
+          tick();
+
+          expect(actualActions).toEqual([
+            actions.stateRehydratedFromUrl({
+              routeKind: RouteKind.COMPARE_EXPERIMENT,
+              partialState: {},
+            }),
+            actions.navigating({
+              after: buildRoute({
+                routeKind: RouteKind.COMPARE_EXPERIMENT,
+                params: {},
+                pathname: '/compare',
+                queryParams: [],
+                navigationOptions: {
+                  replaceState: true,
+                },
+              }),
+            }),
+            jasmine.any(Object),
+          ]);
+        }));
+
+        it('redirects with query parameter', fakeAsync(() => {
+          getPathSpy.and.returnValue('/redirect_query/compare');
+          getSearchSpy.and.returnValue([{key: 'not', value: 'used'}]);
+          // Query paramter formed by the redirector is passed to the
+          // deserializer instead of one from Location.getSearch(). This
+          // behavior emulates redirected URL to be on the URL bar. That is,
+          // when passing through the redirection flow, the actual URL on the
+          // location bar is pre-redirection, "/redirect_query/compare". If
+          // redirector wants to "set" href to "/compare?foo=bar", the query
+          // parameter has to come from the redirector, not from the URL bar.
+          deserializeQueryParamsSpy
+            .withArgs([{key: 'hard', value: 'coded'}])
+            .and.returnValue({good: 'value'})
+            .and.throwError('Invalid');
```

Comment:
I'm not familiar with what happens when a spy has both `.and.returnValue` and `.and.throwError`, does this return a value on the first call, and throw an error on subsequent calls?

Reply:
PTAL at https://stackblitz.com/edit/jasmine-testing-v98jbt?file=src%2Fidentity.spec.ts.

In words, if the method is called with argument that is not specified `withArgs(X).and.returnValue()`, it will throw.